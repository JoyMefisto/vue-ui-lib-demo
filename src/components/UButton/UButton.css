.u-btn {
  display: inline-flex;
  align-items: center;
  justify-content: center;

  font-family: inherit;
  font-weight: 500;
  line-height: 1.2;

  border: none;
  border-radius: var(--btn-radius);
  padding: var(--btn-padding-md);

  box-sizing: border-box;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;

  transition:
    background-color 0.2s ease,
    color 0.2s ease,
    opacity 0.2s ease,
    box-shadow 0.2s ease,
    transform 0.1s ease;
}

/* Sizes */
.u-btn--sm {
  padding: var(--btn-padding-sm);
  font-size: var(--btn-font-size-sm);
}
.u-btn--md {
  padding: var(--btn-padding-md);
  font-size: var(--btn-font-size-md);
}
.u-btn--lg {
  padding: var(--btn-padding-lg);
  font-size: var(--btn-font-size-lg);
}

/* Variants */
.u-btn--primary {
  background-color: var(--btn-primary-bg);
  color: var(--btn-primary-color);
}
.u-btn--secondary {
  background-color: var(--btn-secondary-bg);
  color: var(--btn-secondary-color);
}
.u-btn--warning {
  background-color: var(--btn-warning-bg);
  color: var(--btn-warning-color);
}
.u-btn--danger {
  background-color: var(--btn-danger-bg);
  color: var(--btn-danger-color);
}

/* States */
.u-btn:hover:not(.is-disabled):not(:disabled) {
  box-shadow: var(--btn-active-shadow);
}
.u-btn:active:not(.is-disabled):not(:disabled) {
  transform: scale(var(--btn-active-scale));
}
.u-btn:focus {
  outline: none;
}
.u-btn:focus-visible {
  box-shadow: var(--btn-active-shadow);
}

/* Disabled */
.is-disabled,
.u-btn:disabled {
  opacity: var(--btn-disabled-opacity);
  cursor: not-allowed;
  pointer-events: none;
}
