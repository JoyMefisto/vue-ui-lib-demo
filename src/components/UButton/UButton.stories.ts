import type { Meta, StoryObj } from '@storybook/vue3'
import UButton from './UButton.vue'

const meta: Meta<typeof UButton> = {
  title: 'UI/UButton',
  component: UButton,
  tags: ['autodocs'],
  argTypes: {
    variant: {
      control: { type: 'select' },
      options: ['primary', 'secondary', 'warning', 'danger'],
    },
    size: {
      control: { type: 'select' },
      options: ['sm', 'md', 'lg'],
    },
    disabled: { control: 'boolean' },
    onClick: { action: 'clicked' },
  },
}

export default meta

type Story = StoryObj<typeof UButton>

// 1. Basic usage
export const Primary: Story = {
  args: {
    variant: 'primary',
    size: 'md',
    disabled: false,
    default: 'Primary Button',
  },
}

// 2. All variants
export const Variants: Story = {
  render: () => ({
    components: { UButton },
    template: `
      <div>
        <UButton variant="primary">Primary</UButton>
        <UButton variant="secondary">Secondary</UButton>
        <UButton variant="warning">Warning</UButton>
        <UButton variant="danger">Danger</UButton>
      </div>
    `,
  }),
}

// 3. All sizes
export const Sizes: Story = {
  render: () => ({
    components: { UButton },
    template: `
      <div>
        <UButton size="sm">Small</UButton>
        <UButton size="md">Medium</UButton>
        <UButton size="lg">Large</UButton>
      </div>
    `,
  }),
}

// 4. Disabled state
export const Disabled: Story = {
  render: () => ({
    components: { UButton },
    template: `<UButton disabled>Disabled Button</UButton>`,
  }),
}

// 5. Click handler
export const WithClick: Story = {
  args: {
    default: 'Click Me',
  },
}
