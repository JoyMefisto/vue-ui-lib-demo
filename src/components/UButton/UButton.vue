<template>
  <button
    :class="[baseClass, sizeClass, variantClass]"
    :disabled="disabled"
    @click="$emit('click', $event)"
  >
    <slot />
  </button>
</template>

<script lang="ts" setup>
import '@/styles/theme.css'
import '@/styles/base.css'
import './UButton.css'

import { computed } from 'vue'

export type UButtonSize = 'sm' | 'md' | 'lg'
export type UButtonVariant = 'primary' | 'secondary' | 'warning' | 'danger'

export interface UButtonProps {
  variant?: 'primary' | 'secondary' | 'warning' | 'danger'
  size?: 'sm' | 'md' | 'lg'
  disabled?: boolean
}

const props = withDefaults(defineProps<UButtonProps>(), {
  variant: 'primary',
  size: 'md',
  disabled: false,
})

defineEmits<{
  (e: 'click', ev: MouseEvent): void
}>()

const baseClass = 'u-btn'
const sizeClass = computed(() => `u-btn--${props.size}`)
const variantClass = computed(() => `u-btn--${props.variant}`)
</script>
