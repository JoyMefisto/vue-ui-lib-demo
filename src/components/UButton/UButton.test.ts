import { mount } from '@vue/test-utils'
import { describe, it, expect } from 'vitest'
import UButton from './UButton.vue'

describe('UButton', () => {
  it('renders slot content', () => {
    const wrapper = mount(UButton, {
      slots: { default: 'Click me' },
    })

    expect(wrapper.text()).toBe('Click me')
  })

  it('applies primary variant by default', () => {
    const wrapper = mount(UButton)

    expect(wrapper.classes()).toContain('u-btn--primary')
  })

  it('applies correct variant class', () => {
    const wrapper = mount(UButton, {
      props: { variant: 'danger' },
    })

    expect(wrapper.classes()).toContain('u-btn--danger')
  })

  it('applies correct size class', () => {
    const wrapper = mount(UButton, {
      props: { size: 'lg' },
    })

    expect(wrapper.classes()).toContain('u-btn--lg')
  })

  it('emits click event when clicked', async () => {
    const wrapper = mount(UButton, {
      slots: { default: 'Click' },
    })

    await wrapper.trigger('click')

    expect(wrapper.emitted()).toHaveProperty('click')
    expect(wrapper.emitted('click')?.[0]).toBeTruthy()
  })

  it('does not emit click when disabled', async () => {
    const wrapper = mount(UButton, {
      props: { disabled: true },
    })

    await wrapper.trigger('click')

    expect(wrapper.emitted('click')).toBeUndefined()
  })

  it('has disabled attribute when disabled', () => {
    const wrapper = mount(UButton, {
      props: { disabled: true },
    })

    expect(wrapper.attributes('disabled')).toBeDefined()
  })

  it('matches snapshot', () => {
    const wrapper = mount(UButton, {
      slots: { default: 'Snapshot' },
      props: { variant: 'secondary', size: 'sm' },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
