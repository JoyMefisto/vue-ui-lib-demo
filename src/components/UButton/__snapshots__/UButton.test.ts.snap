// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UButton > matches snapshot 1`] = `"<button class="u-btn u-btn--sm u-btn--secondary">Snapshot</button>"`;
